File: packages/simulation-core/intelligence/prediction-engine.js
/**
 * PREDICTIVE STATE ENGINE
 * Predicts battery depletion, weather transitions, communication latency, and mission progress over time horizons.
 */

import { calculateOneWayDelay, calculateRoundTripDelay } from '../../communication-protocol/index.js';
import { UncertainValue, createPrediction } from '../../shared-types/uncertainty.js';

export class MissionPredictionEngine {
  /**
   * Predicts future simulation state over a specified time horizon in seconds.
   * @param {Object} observation Unified AutonomyObservation
   * @param {number} horizonSeconds Horizon duration in seconds
   * @returns {Object} Structured Predictions
   */
  predict(observation, horizonSeconds = 600) {
    if (!observation || typeof observation !== 'object') {
      return { battery: {}, weather: {}, communication: {}, mission: {} };
    }

    const currentSimTime = observation.environment ? (observation.environment.simulationTime || 0) : 0;
    const futureSimTime = currentSimTime + horizonSeconds;

    // 1. Battery Prediction (Objective 4 energy model)
    const currentBattery = observation.rover ? observation.rover.batteryLevel : 0.94;
    const reserve = observation.constraints ? (observation.constraints.minimumBatteryReserve || 0.15) : 0.15;
    const drainRatePerSec = 0.0001; // deterministic movement/ops rate
    const predictedBattery = Math.max(0.0, currentBattery - (horizonSeconds * drainRatePerSec));
    const reserveViolationExpected = predictedBattery < reserve;
    const batteryStd = Math.round((0.02 + (horizonSeconds / 3600.0) * 0.03) * 1000) / 1000;

    const batteryPrediction = {
      currentBattery,
      predictedBattery: Math.round(predictedBattery * 1000) / 1000,
      horizonSeconds,
      reserveViolationExpected,
      confidence: 0.90,
      uncertainty: new UncertainValue(Math.round(predictedBattery * 1000) / 1000, batteryStd, 0.90, 'fraction').toJSON()
    };

    // 2. Weather Prediction (Reusing Objective 4 deterministic weather timeline)
    let predictedWeatherState = 'CLEAR';
    let predictedVisibility = 100;
    let weatherConfidence = 0.95;

    if (futureSimTime >= 7200 && futureSimTime < 10800) {
      predictedWeatherState = 'DUST_STORM';
      predictedVisibility = 10;
      weatherConfidence = 0.92;
    } else if (futureSimTime >= 3600 && futureSimTime < 7200) {
      predictedWeatherState = 'DUSTY';
      predictedVisibility = 50;
      weatherConfidence = 0.90;
    } else {
      predictedWeatherState = 'CLEAR';
      predictedVisibility = 100;
      weatherConfidence = 0.95;
    }

    const weatherPrediction = {
      currentState: observation.environment && observation.environment.weather ? observation.environment.weather.state : 'CLEAR',
      predictedState: predictedWeatherState,
      predictedVisibility,
      horizonSeconds,
      confidence: weatherConfidence,
      uncertainty: new UncertainValue(predictedVisibility, 5.0, weatherConfidence, 'meters').toJSON()
    };

    // 3. Communication Prediction (Reusing Objective 1 & 2 delay engine)
    const distKm = observation.communication ? (observation.communication.distanceKm || 225000000) : 225000000;
    const oneWayDelay = calculateOneWayDelay(distKm);
    const roundTripDelay = calculateRoundTripDelay(distKm);
    const commState = observation.communication ? observation.communication.communicationState : 'AVAILABLE';

    const communicationPrediction = {
      communicationState: commState,
      distanceKm: distKm,
      estimatedOneWayDelay: oneWayDelay,
      estimatedRoundTripDelay: roundTripDelay,
      blackoutExpected: commState === 'BLACKOUT',
      confidence: 0.98,
      uncertainty: new UncertainValue(oneWayDelay, 0.5, 0.98, 'seconds').toJSON()
    };

    // 4. Mission Progress Prediction
    const currentTask = observation.currentTask ? observation.currentTask.name : 'None';
    const currentProgress = observation.missionProgress || 0.0;
    const predictedProgress = Math.min(100.0, currentProgress + 14.28); // +1 task completion

    const missionPrediction = {
      currentTask,
      currentProgress,
      predictedProgress: Math.round(predictedProgress * 100) / 100,
      resourceMarginFeasible: !reserveViolationExpected,
      confidence: 0.88,
      uncertainty: new UncertainValue(Math.round(predictedProgress * 100) / 100, 2.5, 0.88, 'percent').toJSON()
    };

    return {
      battery: batteryPrediction,
      weather: weatherPrediction,
      communication: communicationPrediction,
      mission: missionPrediction
    };
  }
}

File: services/ai-engine/app/intelligence/prediction_engine.py
from app.models.observations import AutonomyObservationModel
from app.models.uncertainty import UncertainValueModel
from app.models.predictions import (
    PredictionsModel,
    BatteryPredictionModel,
    WeatherPredictionModel,
    CommunicationPredictionModel,
    MissionPredictionModel
)

class PyPredictionEngine:
    def predict(self, obs: AutonomyObservationModel, horizon_seconds: float = 600.0) -> PredictionsModel:
        curr_sim_time = obs.environment.simulationTime
        future_sim_time = curr_sim_time + horizon_seconds

        # 1. Battery Trajectory Prediction with Uncertainty
        current_battery = obs.rover.batteryLevel
        reserve = obs.constraints.minimumBatteryReserve
        drain_rate = 0.0001 # per second
        predicted_battery = max(0.0, current_battery - (horizon_seconds * drain_rate))
        battery_std = round(0.02 + (horizon_seconds / 3600.0) * 0.03, 3)
        reserve_violation = predicted_battery < reserve

        battery_pred = BatteryPredictionModel(
            currentBattery=current_battery,
            predictedBattery=round(predicted_battery, 3),
            horizonSeconds=horizon_seconds,
            reserveViolationExpected=reserve_violation,
            confidence=0.90,
            uncertainty=UncertainValueModel.create(mean=round(predicted_battery, 3), std=battery_std, confidence=0.90, unit="fraction")
        )

        # 2. Weather State Prediction with Uncertainty
        if 7200.0 <= future_sim_time < 10800.0:
            pred_weather = "DUST_STORM"
            pred_vis = 10.0
            weather_conf = 0.92
        elif 3600.0 <= future_sim_time < 7200.0:
            pred_weather = "DUSTY"
            pred_vis = 50.0
            weather_conf = 0.90
        else:
            pred_weather = "CLEAR"
            pred_vis = 100.0
            weather_conf = 0.95

        weather_pred = WeatherPredictionModel(
            currentState=obs.environment.weather.state,
            predictedState=pred_weather,
            predictedVisibility=pred_vis,
            horizonSeconds=horizon_seconds,
            confidence=weather_conf,
            uncertainty=UncertainValueModel.create(mean=pred_vis, std=5.0, confidence=weather_conf, unit="meters")
        )

        # 3. Communication Prediction with Uncertainty
        comm_pred = CommunicationPredictionModel(
            communicationState=obs.communication.communicationState,
            distanceKm=obs.communication.distanceKm,
            estimatedOneWayDelay=obs.communication.estimatedOneWayDelay,
            estimatedRoundTripDelay=obs.communication.estimatedRoundTripDelay,
            blackoutExpected=(obs.communication.communicationState == "BLACKOUT"),
            confidence=0.98,
            uncertainty=UncertainValueModel.create(mean=obs.communication.estimatedOneWayDelay, std=0.5, confidence=0.98, unit="seconds")
        )

        # 4. Mission Progress Prediction with Uncertainty
        curr_task = obs.currentTask.get("name", "None") if obs.currentTask else "None"
        curr_progress = obs.missionProgress
        pred_progress = min(100.0, curr_progress + 14.28)

        mission_pred = MissionPredictionModel(
            currentTask=curr_task,
            currentProgress=curr_progress,
            predictedProgress=round(pred_progress, 2),
            resourceMarginFeasible=not reserve_violation,
            confidence=0.88,
            uncertainty=UncertainValueModel.create(mean=round(pred_progress, 2), std=2.5, confidence=0.88, unit="percent")
        )

        return PredictionsModel(
            battery=battery_pred,
            weather=weather_pred,
            communication=comm_pred,
            mission=mission_pred
        )
